src/Routing/PushState.purs

```diff
@@ -0,0 +1,241 @@
+module Routing.PushState
+  ( PushStateEffects
+  , PushStateInterface
+  , LocationState
+  , Basename(..)
+  , Path(..)
+  , makeInterface
+  , foldLocations
+  , locations
+  , foldPaths
+  , paths
+  , matches
+  , matchesWith
+  ) where
+
+import Prelude
+
+import Control.Monad.Eff (Eff)
+import Control.Monad.Eff.Ref (REF, modifyRef, newRef, readRef, writeRef)
+import DOM (DOM)
+import DOM.Event.EventTarget (addEventListener, eventListener) as DOM
+import DOM.HTML (window) as DOM
+import DOM.HTML.Event.EventTypes (popstate) as DOM
+import DOM.HTML.History (DocumentTitle(..), URL(..), pushState, replaceState, state) as DOM
+import DOM.HTML.Location (hash, pathname, search) as DOM
+import DOM.HTML.Types (HISTORY)
+import DOM.HTML.Types (htmlDocumentToDocument, windowToEventTarget) as DOM
+import DOM.HTML.Window (document, history, location) as DOM
+import DOM.Node.Document (createTextNode) as DOM
+import DOM.Node.MutationObserver (mutationObserver, observe) as DOM
+import DOM.Node.Node (setNodeValue) as DOM
+import DOM.Node.Types (textToNode) as DOM
+import Data.Array as Array
+import Data.Foldable (class Foldable, for_, indexl, traverse_)
+import Data.Foreign (Foreign)
+import Data.Maybe (Maybe(..), fromMaybe, maybe)
+import Data.Newtype (class Newtype, unwrap)
+import Data.StrMap as StrMap
+import Data.String as String
+import Routing (match)
+import Routing.Match (Match)
+
+type PushStateEffects eff =
+  ( history :: HISTORY
+  , dom :: DOM
+  , ref :: REF
+  | eff
+  )
+
+newtype Basename = Basename String
+
+derive newtype instance eqBasename :: Eq Basename
+derive newtype instance ordBasename :: Ord Basename
+derive instance newtypeBasename :: Newtype Basename _
+
+newtype Path = Path String
+
+derive newtype instance eqPath :: Eq Path
+derive newtype instance ordPath :: Ord Path
+derive instance newtypePath :: Newtype Path _
+
+-- | A `PushStateInterface` is a localized instance for making location changes
+-- | and consuming the events. Since the DOM API does not provide a general
+-- | event type for push state (only pop), you must use the coupled effects
+-- | along with `listen` to receive events on all location changes.
```

Comment:
So, if I have two PureScript apps on the same web page, both using this pushstate routing lib, and each of them pushes a new route using pushState, their instances of `PushStateInterface` will represent an incorrect view of the browser's history state, because they can't see see each other's pushState events?

Reply:
It won't represent an incorrect view so much as different instances won't get notified on push/replace. Dereferencing the state is always correct. I don't know how to solve this without module level unsafe effects and global state, which I'm not going to do. I am open to suggestions though.